pkg/controller/bastion/actuator.go

```diff
@@ -0,0 +1,205 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bastion
+
+import (
+	"context"
+	"encoding/json"
+
+	api "github.com/gardener/gardener-extension-provider-azure/pkg/apis/azure"
+	azureclient "github.com/gardener/gardener-extension-provider-azure/pkg/azure/client"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2021-03-01/compute"
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2020-05-01/network"
+	"github.com/gardener/gardener/extensions/pkg/controller"
+	"github.com/gardener/gardener/extensions/pkg/controller/bastion"
+	"github.com/go-logr/logr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type actuator struct {
+	client client.Client
+	logger logr.Logger
+}
+
+func newActuator() bastion.Actuator {
+	return &actuator{
+		logger: logger,
+	}
+}
+
+func (a *actuator) InjectClient(client client.Client) error {
+	a.client = client
+	return nil
+}
+
+func createBastionInstance(ctx context.Context, factory azureclient.Factory, opt *Options, parameters *compute.VirtualMachine) (*compute.VirtualMachine, error) {
+	vmClient, err := factory.VirtualMachine(ctx, opt.SecretReference)
+	if err != nil {
+		return nil, err
+	}
+
+	instance, err := vmClient.Create(ctx, opt.ResourceGroupName, opt.BastionInstanceName, parameters)
+	if err != nil {
+		logger.Info("Unable to Create VM Instance,", "instance_name", opt.BastionInstanceName)
+		return nil, err
+	}
+	return instance, nil
+}
+
+func createPublicIP(ctx context.Context, factory azureclient.Factory, opt *Options, parameters *network.PublicIPAddress) (*network.PublicIPAddress, error) {
+	publicClient, err := factory.NetworkPublicIP(ctx, opt.SecretReference)
+	if err != nil {
+		return nil, err
+	}
+
+	ip, err := publicClient.CreateOrUpdate(ctx, opt.ResourceGroupName, opt.BastionPublicIPName, *parameters)
+	if err != nil {
+		logger.Info("Unable to Create Public IP address,", "public_ip_address_name", opt.BastionPublicIPName)
+		return nil, err
+	}
+	return ip, nil
+}
+
+func createOrUpdateNetworkSecGroup(ctx context.Context, factory azureclient.Factory, opt *Options, parameters *network.SecurityGroup) error {
+	nsgClient, err := factory.NetworkSecurityGroup(ctx, opt.SecretReference)
+	if err != nil {
+		return err
+	}
+
+	_, err = nsgClient.CreateOrUpdate(ctx, opt.ResourceGroupName, opt.SecurityGroupName, *parameters)
+	if err != nil {
+		logger.Error(err, "Network Security Group cannot be update: ", "securityGroup", opt.SecurityGroupName)
```

Comment:
fix https://github.com/gardener/gardener-extension-provider-azure/pull/397/commits/93cd40d5bc49f99da42d0c7181db617291f34c69

Reply:
@petersutter 
answer for 3: every time we create a bastion instance we use a unique name for it which is generated (hashed) at `generateBastionBaseResourceName`. And every time we create an individual network security rule and delete it afterward.

Reply:
regarding 3: @andrei-panov I was talking about the network security group and not the network security rule

Reply:
I tested individual manipulation of SecurityRules and have got this error: `does not have authorization to perform action`.
Together with @petersutter we agreed to address this as a separate PR because it's already many changes and it's hard to follow. The current implementation is fully functional.